File: handlers/system.rb
# frozen_string_literal: true

require "base64"
require "tmpdir"

module TuSketchupAgent
  module Handlers
    module System
      extend self

      def register
        Router.register("ping") { ping_response }
        Router.register("model_summary") { model_summary }
        Router.register("get_selection") { get_selection }
        Router.register("zoom_extents") { zoom_extents }
        Router.register("capture_viewport") { |args| capture_viewport(args) }
        Router.register("execute_ruby") { |args| execute_ruby(args) }
      end

      def ping_response
        {
          ok: true,
          service: "tu-sketchup-agent",
          protocol_version: TuSketchupAgent::PROTOCOL_VERSION,
          min_compatible_protocol_version: TuSketchupAgent::MIN_COMPATIBLE_PROTOCOL_VERSION,
          sketchup_version: Sketchup.version,
          server: "#{TuSketchupAgent::HOST}:#{TuSketchupAgent::PORT}",
          dev_mode: Auth.dev_mode?
        }
      end

      def model_summary
        model = Sketchup.active_model
        raise "Không có model nào đang mở" unless model

        bbox = model.bounds
        {
          ok: true,
          service: "tu-sketchup-agent",
          protocol_version: TuSketchupAgent::PROTOCOL_VERSION,
          min_compatible_protocol_version: TuSketchupAgent::MIN_COMPATIBLE_PROTOCOL_VERSION,
          sketchup_version: Sketchup.version,
          model_revision: Operation.model_revision,
          mcp_revision: Operation.model_revision,
          model_revision_source: "tu-sketchup-agent",
          capabilities: {
            find_entity_by_persistent_id: model.respond_to?(:find_entity_by_persistent_id),
            find_entity_by_id: model.respond_to?(:find_entity_by_id)
          },
          title: model.title.empty? ? "Untitled" : model.title,
          path: model.path,
          entity_count: model.active_entities.length,
          selection_count: model.selection.length,
          layers: model.layers.map(&:name),
          materials: model.materials.map(&:name),
          scenes: model.pages.map(&:name),
          bounds_mm: {
            width: bbox.width.to_mm.round(1),
            depth: bbox.height.to_mm.round(1),
            height: bbox.depth.to_mm.round(1)
          }
        }
      end

      def get_selection
        model = Sketchup.active_model
        raise "Không có model nào đang mở" unless model

        items = model.selection.map do |entity|
          bbox = entity.respond_to?(:bounds) ? entity.bounds : nil
          meta = Services::MetadataService.entity_metadata(entity)
          meta.merge(
            material: entity.respond_to?(:material) && entity.material ? entity.material.name : nil,
            bounds_mm: bbox ? {
              width: bbox.width.to_mm.round(1),
              depth: bbox.height.to_mm.round(1),
              height: bbox.depth.to_mm.round(1),
              center: [bbox.center.x.to_mm.round(1), bbox.center.y.to_mm.round(1), bbox.center.z.to_mm.round(1)]
            } : nil
          )
        end

        { ok: true, count: items.length, items: items }
      end

      def zoom_extents
        model = Sketchup.active_model
        raise "Không có model nào đang mở" unless model

        model.active_view.zoom_extents
        { ok: true, operation: "zoom_extents" }
      end

      def capture_viewport(args)
        width = (args["width"] || 1280).to_i
        height = (args["height"] || 720).to_i
        raise ArgumentError, "width phải trong khoảng 64..4096" unless width.between?(64, 4096)
        raise ArgumentError, "height phải trong khoảng 64..4096" unless height.between?(64, 4096)

        output_path = args["output_path"].to_s.strip
        is_temp = output_path.empty?
        target_file = is_temp ? File.join(Dir.tmpdir, "sketchup_vp_#{Time.now.to_i}_#{rand(1000)}.png") : output_path

        include_base64 = args.fetch("include_base64", false)

        model = Sketchup.active_model
        raise "Không có model nào đang mở" unless model

        view = model.active_view

        view.write_image({
          filename: target_file,
          width: width,
          height: height,
          antialias: true,
          compression: 0.9
        })

        image_written = File.exist?(target_file)
        b64_data = (include_base64 && image_written) ? Base64.strict_encode64(File.binread(target_file)) : nil

        {
          ok: true,
          width: width,
          height: height,
          image_path: is_temp ? nil : target_file,
          image_available: image_written,
          image_base64: b64_data
        }
      ensure
        File.delete(target_file) if is_temp && target_file && File.exist?(target_file)
      end

      def execute_ruby(args)
        unless Auth.dev_mode?
          return {
            ok: false,
            error: {
              code: "DEV_MODE_REQUIRED",
              message: "execute_ruby bị vô hiệu hóa ngoài dev mode (bật TU_SKETCHUP_DEV_MODE=1 hoặc qua menu Extension)",
              class: "SecurityError"
            }
          }
        end

        code = args["code"].to_s
        if code.strip.empty?
          return {
            ok: false,
            error: {
              code: "INVALID_ARGUMENT",
              message: "Thiếu code ruby"
            }
          }
        end

        if code.bytesize > 100_000
          return {
            ok: false,
            error: {
              code: "PAYLOAD_TOO_LARGE",
              message: "Code vượt quá giới hạn 100 KB"
            }
          }
        end

        auto_op = args.fetch("auto_operation", false)
        model = Sketchup.active_model

        puts "TuSketchupAgent: [execute_ruby] #{code.slice(0, 80).strip}..."

        result = nil
        if auto_op && model
          Operation.with_operation(model, "AI Ruby Execution") do
            result = eval(code, TOPLEVEL_BINDING)
            Operation.bump_model_revision
          end
        else
          result = eval(code, TOPLEVEL_BINDING)
        end

        {
          ok: true,
          result: result.inspect,
          class: result.class.name,
          model_revision: Operation.model_revision,
          mcp_revision: Operation.model_revision
        }
      rescue StandardError => error
        err = {
          code: "RUBY_EXECUTION_ERROR",
          message: error.message,
          class: error.class.name
        }
        err[:backtrace] = error.backtrace.first(5) if Auth.dev_mode? && error.backtrace
        {
          ok: false,
          error: err
        }
      end
    end
  end

  # Module-level delegates for backward compatibility
  def self.ping_response
    Handlers::System.ping_response
  end

  def self.model_summary
    Handlers::System.model_summary
  end

  def self.get_selection
    Handlers::System.get_selection
  end

  def self.zoom_extents
    Handlers::System.zoom_extents
  end

  def self.capture_viewport(args)
    Handlers::System.capture_viewport(args)
  end

  def self.execute_ruby(args)
    Handlers::System.execute_ruby(args)
  end
end

File: mcp_agent/tools/system.py
"""
System & Connection Tools for Tu SketchUp Agent.
"""

import json
import base64
from typing import Optional
from mcp.server.fastmcp import FastMCP, Image
from ..transport import send_to_sketchup


def register(mcp: FastMCP) -> None:
    @mcp.tool()
    def sketchup_ping() -> str:
        """Kiểm tra trạng thái kết nối tới SketchUp 2026."""
        res = send_to_sketchup("ping")
        return json.dumps(res, indent=2, ensure_ascii=False)

    @mcp.tool()
    def sketchup_get_model_info() -> str:
        """Lấy thông tin tổng quan của mô hình SketchUp đang mở (kích thước, layers/tags, materials, scenes, số lượng entity)."""
        res = send_to_sketchup("model_summary")
        return json.dumps(res, indent=2, ensure_ascii=False)

    @mcp.tool()
    def sketchup_get_selection() -> str:
        """Xem danh sách và chi tiết các đối tượng (entity) đang được chọn trong SketchUp."""
        res = send_to_sketchup("get_selection")
        return json.dumps(res, indent=2, ensure_ascii=False)

    @mcp.tool()
    def sketchup_zoom_extents() -> str:
        """Zoom toàn cảnh viewport để bao trọn tất cả các hình khối trong model."""
        res = send_to_sketchup("zoom_extents")
        return json.dumps(res, indent=2, ensure_ascii=False)

    @mcp.tool()
    def sketchup_capture_viewport(width: int = 1280, height: int = 720) -> Image:
        """
        Chụp ảnh góc nhìn 3D hiện tại trong SketchUp Viewport và trả về ảnh để AI quan sát kết quả.
        
        Args:
            width: Chiều rộng ảnh (mặc định 1280px)
            height: Chiều cao ảnh (mặc định 720px)
        """
        if not 64 <= width <= 4096:
            raise ValueError("width phải trong khoảng 64..4096")

        if not 64 <= height <= 4096:
            raise ValueError("height phải trong khoảng 64..4096")

        res = send_to_sketchup("capture_viewport", {"width": width, "height": height, "include_base64": True})
        if not res.get("ok"):
            err = res.get("error", {})
            err_msg = err.get("message") if isinstance(err, dict) else str(err)
            raise RuntimeError(f"Lỗi chụp viewport SketchUp: {err_msg}")

        b64_data = res.get("image_base64")
        if not b64_data:
            raise RuntimeError(f"Không nhận được dữ liệu hình ảnh từ SketchUp: {res}")

        img_bytes = base64.b64decode(b64_data)
        return Image(data=img_bytes, format="png")

    @mcp.tool()
    def sketchup_execute_ruby(code: str, auto_operation: bool = False) -> str:
        """
        Thực thi mã Ruby trong SketchUp Ruby API.
        
        Lưu ý bảo mật:
        - Yêu cầu Dev Mode phải được BẬT trực tiếp trong SketchUp (qua menu Extensions -> Tu SketchUp Agent -> Toggle Dev Mode hoặc biến môi trường TU_SKETCHUP_DEV_MODE=1).
        - Không thể tự bật Dev Mode qua mạng để đảm bảo an toàn tuyệt đối cho người dùng.
        - Mặc định auto_operation=False để script tự quyết định transaction hoặc chạy lệnh tra cứu read-only mà không tạo Undo rác.
        - Giới hạn kích thước code: 100 KB.
        """
        res = send_to_sketchup("execute_ruby", {"code": code, "auto_operation": auto_operation}, timeout=30.0)
        return json.dumps(res, indent=2, ensure_ascii=False)
